app/lib/link_details_extractor.rb

```diff
@@ -269,14 +269,31 @@ def structured_data
   end
 
   def document
-    @document ||= Nokogiri::HTML(@html, nil, encoding)
+    @document ||= begin
+      encoding = detect_encoding || parse_encoding || header_encoding || 'UTF-8'
+      @document || Nokogiri::HTML(@html, nil, encoding)
```

Comment:
I think I get why this is like this (avoid the double parse) ... but its slightly unexpected, especially inside the memoizing block assigning ... itself. I can't think of a better way to avoid double parse w/out getting a lot more verbose (ie, each encoding option method could return nil or a fully parsed doc if its attempt was succesful?)

Not a blocker, just wanted to note. Maybe future refactor.

Reply:
Would it make more sense to go with this pattern instead?

```ruby
def document
  return @document if defined?(@document)
  encoding = detect_encoding || parse_encoding || header_encoding || 'UTF-8'
  @document = Nokogiri::HTML(@html, nil, encoding)
end
```

Or maybe `encoding` could be its own method (memoized?)

Reply:
This pattern wouldn't help with avoiding the double-parsing. The current code of this PR is because `parse_encoding` actually assigns `@document`.

Taking a step back, `LinkDetailsExtractor#encoding` and `@encoding` are only used to correctly parse the document, so this can be factored in a complete different way if needs be.

Reply:
I am not against being more verbose here if it helps readability. I played around a bit and came up with the following, which is not super elegant but hopefully a lot less surprising:

```ruby
def document
    @document ||= detect_encoding_and_parse_document
  end

  def detect_encoding_and_parse_document
    encoding = detect_encoding
    encoding, document = parse_encoding if encoding.nil?
    encoding ||= header_encoding
    encoding ||= 'UTF-8'
    document || Nokogiri::HTML(@html, nil, encoding)
  end

  def detect_encoding
    guess = detector.detect(@html, @html_charset)
    guess&.fetch(:confidence, 0).to_i > 60 ? guess&.fetch(:encoding, nil) : nil
  end

  def parse_encoding
    parsed_document = Nokogiri::HTML(@html)
    source_with_detected_encoding = parsed_document.to_s
    encoding = source_with_detected_encoding.encoding.name if source_with_detected_encoding.valid_encoding?
    [encoding, encoding ? parsed_document : nil]
  end

  def header_encoding
    return unless @html_charset

    @html.force_encoding(@html_charset)
    @html_charset if @html.valid_encoding?
  end
```

wdyt?

Reply:
Maybe `detect_encoding_and_parse_document` could be rewritten as:

```ruby
  def detect_encoding_and_parse_document
    encoding = detect_encoding
    return Nokogiri::HTML(@html, nil, encoding) if encoding.present?

    parse_without_encoding || Nokogiri::HTML(@html, nil, header_encoding || 'UTF-8')
  end
```

And `parse_encoding` replaced with:
```ruby
def parse_without_encoding
  parsed_document = Nokogiri::HTML(@html)
  return unless parsed_document.to_s.valid_encoding?

  parsed_document
end
```

Alternatively, checking `valid_encoding?` is something we'd always do, with something like (untested):
```
def detect_encoding_and_parse_document
  [detect_encoding, nil, header_encoding, 'UTF-8'].uniq.each do |encoding|
    document = Nokogiri::HTML(@html, nil, encoding)
    return document if document.to_s.valid_encoding?
  end
end
```

Reply:
I love that last version. In this case we can even ditch `header_encoding`, so this is the shortest solution yet, while at the same time avoiding the confusion mentioned above. Thank you so much @ClearlyClaire !